man/tcosmonitor.1
.TH tcosmonitor 1 "Mar 11, 2007" "tcosmonitor man page"

.SH NAME
tcosmonitor \- Python GTK2 gui that allow an admin or teacher to take control of a thin client classrom.

.SH DESCRIPTION

.PP
\fBtcosmonitor\fP is a python complete GUI with which a teacher or admin can :
 * Reboot / Poweroff all connected host or single one.
 * Lock / Unlock screens.
 * Take a screenshot
 * Send text, exec app, kill app
 * etc...

TcosMonitor uses auth with tcosxmlrpc (user and password)

.SH OPTIONS
.B [no options]
 With no options will run in normal mode.

.B --help
 Some usage help about tcosmonitor

.B --debug
 Show verbose output when running

.SH BUGS
Contact Mario Izquierdo <mariodebian@gmail.com>

Homepage: http://www.tcosproject.org

.SH SEE ALSO
tcosmonitor(1)

.SH AUTHOR
Mario Izquierdo (mariodebian) <mariodebian@gmail.com>


Changelog
tcosmonitor (0.1.0) unstable; urgency=low

  * Not build LTSP or PXES packages anymore (not clean, not debian)
  * Not get busybox-static
  * Use python-support to bytecompile packages
  * Update debhelper Build-dep
  * Move TcosMonitor desktop icon to Admin menu
  * Hide tcospersonalize menu
  * Not ignore debhelper in tcos-tcosmonitor.postinst
  * Change tcospersonalize config dir to new /var/lib/tcos/tftp/config

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sun, 18 Feb 2007 13:36:53 +0100

tcosmonitor (0.0.19.1) unstable; urgency=low

  * xmlrpc/*:
     - Change #ifdef DEBUG for a new function dbgtcos(); read TCOS_DEBUG env var

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Thu,  8 Feb 2007 18:33:04 +0100

tcosmonitor (0.0.18.2) unstable; urgency=low

  * Downgrade initramfs-tools-tcos dependencies to >= 0.59
    tcosmonitor must be installed before initramfs-tools-tcos 0.60

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sun,  4 Feb 2007 22:04:25 +0100

tcosmonitor (0.0.18.1) unstable; urgency=low

  * hooks-addons/tcosmonitor:
     - remove copy od seq (fixed in initramfs-tools-tcos 0.60)
  * debian/control:
     - update version depends
  * Remove hex2ascii bin => initramfs-tools-tcos

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat,  3 Feb 2007 20:39:01 +0100

tcosmonitor (0.0.17.1) unstable; urgency=low

  * xmlrpc/*.c:
     - Remove tcosxmlrpc verbose messages
  * debian/control, hooks-addons/tcosmonitor:
     - Add /usr/bin/eject and eject cdrom from TcosDevices (daemon)
  * tcos-devices.py:
     - Remove sys.exit(1) from tcos-devices (retry cookie don't match)
  * xmlrpc/sh/devicesctl.sh
     - Add --eject method

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon, 29 Jan 2007 16:13:43 +0100

tcosmonitor (0.0.16.2) unstable; urgency=low

  * Change /bin/sh to /bin/bash (solve ubuntu dash problems)

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon, 29 Jan 2007 16:02:30 +0100

tcosmonitor (0.0.16.1) unstable; urgency=low

  * New stable release.
  * Add vertical label in all screenshots.
  * Add uptime var (readed in TCOS general info)
  * Change preferences minicrot_size default (20) and step (1)
  * Change version string in shared.py (in future read from debian/changelog when building package)
  * Fix typo in tcos-server-utils.py

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Tue, 26 Dec 2006 23:35:11 +0100

tcosmonitor (0.0.15.14) unstable; urgency=low

  * Add New toolbar button (All Hosts) that show all_actions_menu
  * Add new global action: "Capture all hosts screenshots"
  * New Preferences var: miniscrot_size (default 15%)
  * tcos-volume-manager:
     - add refresh button to reload all sliders values.

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 23 Dec 2006 21:10:31 +0100

tcosmonitor (0.0.15.13) unstable; urgency=low

  * Fix typo on on_another_screenshot() Worker => Workers

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon, 11 Dec 2006 16:12:48 +0100

tcosmonitor (0.0.15.12) unstable; urgency=low

  * Add set_sensitive() to TreeView before/after updating list or 
    connecting to thin client.
  * Fix app/msg dialog, print hostname/username in new lines.

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon,  4 Dec 2006 09:46:20 +0100

tcosmonitor (0.0.15.11) unstable; urgency=low

  * Fix LocalData.py who |grep "%s "  (add space after %s)

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 29 Nov 2006 19:07:45 +0100

tcosmonitor (0.0.15.10) unstable; urgency=low

  * Put 1d if hours connected < 0 (connected yesterday)

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 29 Nov 2006 18:46:20 +0100

tcosmonitor (0.0.15.9) unstable; urgency=low

  * Fix minor problem in import headers in shared.py file.
  * Fix version retrieve error (disable split in mayor, med and minor)
  * Update web and translation.

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Tue, 28 Nov 2006 21:36:10 +0100

tcosmonitor (0.0.15.8) unstable; urgency=low

  * xmlrpc/Makefile
          -DVERSION=\"fooo\" not use "'foo'"
  * LocalData.py
       Add print_debug in version get
  * shared.py Add some hidden process to list
       ( xauth, tcos-devices, tcos-volume-manager, notification-daemon, bash )

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sun, 19 Nov 2006 20:14:59 +0100

tcosmonitor (0.0.15.7) unstable; urgency=low

  * allow to enable/disable tcos-devices and tcos-volume-manager
    from a new conf file: /etc/tcos/tcosmonitor.conf in runtime.

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed,  8 Nov 2006 13:11:41 +0100

tcosmonitor (0.0.15.6) unstable; urgency=low

  * tcos-volume-manager.py, tcos-volume-manager.glade, tcos-volume-manager.desktop
   - iconize in systray tcos-volume-manager
   - new systray icon for tcos-volume-manager
   - add xdg/autostart tcos-volume-manager (desktop file)
   - exit if xauth dont match
   - mainwindow hide by default in glade file

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 28 Oct 2006 15:38:06 +0200

tcosmonitor (0.0.15.5) unstable; urgency=low

  * tcos-devices.py
    - fix get_value(), DEVICE is not a udev var, varname is DEVNAME
    - fix get_local_mountpoint() fslabel and fsvendor

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 28 Oct 2006 11:54:13 +0200

tcosmonitor (0.0.15.4) unstable; urgency=low

  * tcos-devices.py, udev/tcos-udevd.sh , udev/050_tcos_devices.rules
    - Change udev to control all SUBSYSTEM=="block" events
    - Delete duplicate events, kernel do 4 umount and 1 mount event for mounting a floppy

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 25 Oct 2006 17:24:14 +0200

tcosmonitor (0.0.15.3) unstable; urgency=low

  * TcosPersonalize: run as root only
  * TcosActions: fix update info bug
  * TcosXmlRpc, TcosXauth, TcosConf, ping: import gtk modules if env var DISPLAY is set


tcosmonitor.py
#!/usr/bin/env python2.4
# -*- coding: UTF-8 -*-
##########################################################################
# TcosMonitor writen by MarioDebian <mariodebian@gmail.com>
#
#    TcosMonitor version 0.0.15
#
# Copyright (c) 2006 Mario Izquierdo <mariodebian@gmail.com>
# All rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
###########################################################################

import sys
import os

if not os.path.isfile("Initialize.py"):
	#print "DEBUG: append tcosmonitor dir"
	sys.path.append("/usr/share/tcosmonitor")

import pygtk
pygtk.require('2.0')
from gtk import *
import gtk.glade

from time import time, sleep
import getopt
from gettext import gettext as _
from subprocess import Popen, PIPE, STDOUT
import popen2
from threading import Thread
#import threading

# deprecated
#gtk.threads_init()
gtk.gdk.threads_init()

import gobject

import shared




def print_debug(txt):
    if shared.debug:
        print "%s::%s" %("tcosmonitor", txt)
    return

def crono(start, txt):
    print_debug ("crono(), %s get %f seconds" %(txt, (time() - float(start))) )
    return

def usage():
    print "TcosMonitor help:"
    print ""
    print "   tcosmonitor -d [--debug]  (write debug data to stdout)"
    print "   tcosmonitor -h [--help]   (this help)"


try:
    opts, args = getopt.getopt(sys.argv[1:], ":hd", ["help", "debug"])
except getopt.error, msg:
    print msg
    print "for command line options use tcosconfig --help"
    sys.exit(2)

# process options
for o, a in opts:
    if o in ("-d", "--debug"):
        print "DEBUG ACTIVE"
        shared.debug = True
    if o in ("-h", "--help"):
        usage()
        sys.exit()



import Initialize
import TcosXmlRpc
import LocalData
import TcosConf
import TcosActions
import TcosXauth

class TcosMonitor:
    def __init__(self):
        # if true auto-update is active, false only one update        
        self.updating=False
        self.name="TcosMonitor"
        
        self.worker_running=False
        
        #import shared
        gtk.glade.bindtextdomain(shared.PACKAGE, shared.LOCALE_DIR)
        gtk.glade.textdomain(shared.PACKAGE)
        
        # Widgets
        self.ui = gtk.glade.XML(shared.GLADE_DIR + 'tcosmonitor.glade')
        self.mainwindow = self.ui.get_widget('mainwindow')
        self.mainwindow.set_icon_from_file(shared.IMG_DIR +\
                                     'tcos-icon-32x32.png')
        #self.pref = self.ui.get_widget('prefwindow')
        #self.main.fullscreen()
        self.is_fullscreen=False
        
        # close windows signals
        self.mainwindow.connect('destroy', self.salirse )
        #self.pref.connect('destroy', self.prefwindow_close )
        #self.pref.hide()
        
        
        # FIXME
        self.scrolledtextview = self.ui.get_widget('scrolledtextview')
        import htmltextview
        #htmltextview.HtmlHandler().set_main(self)
        
        self.datatxt = htmltextview.HtmlTextView(self)
        self.datatxt.show()
        self.scrolledtextview.add(self.datatxt)
        self.datatxt.clean()
        
        
        
        # init classes
        self.config=TcosConf.TcosConf(self)
        self.localdata=LocalData.LocalData(self)
        self.xmlrpc=TcosXmlRpc.TcosXmlRpc(self)
        self.xauth=TcosXauth.TcosXauth(self)
        
        
        self.init=Initialize.Initialize(self)
        self.actions=TcosActions.TcosActions(self)
        
        
        
        #########  init some elements ###########
        self.init.init_progressbar()
        self.init.initabout()
        self.init.initask()
        self.init.initpref()
        self.init.populate_pref()
        #########################################
        
        self.actions.update_hostlist()
        self.init.initbuttons()
        
        # make menus
        self.actions.RightClickMenuOne(None)
        self.actions.RightClickMenuAll()
        
        # init hostlist
        self.init.init_hostlist()
        
        if not shared.dbus_disabled:
            from TcosDBus import TcosDBusAction
            self.dbus_action=TcosDBusAction(admin=self.config.GetVar("xmlrpc_username"),
                                  passwd=self.config.GetVar("xmlrpc_password")  )
        
        
        # generate host list if checked
        if self.config.GetVar("populate_list_at_startup") == "1":
            self.populate_host_list()
        

        
    
        
    def prefwindow_close(self, widget, event):
        print_debug ( "prefwindow_close() closing preferences window" )
        self.pref.hide()
        return True
    


    

    def search_host(self, widget):
        print_debug ( "search_host()" )
        txt=self.searchtxt.get_text()
        if txt == "":
            allclients=self.localdata.GetAllClients( self.config.GetVar("scan_network_method") )
            Thread( target=self.actions.populate_hostlist, args=([allclients]) ).start()
        #print txt
        model=self.tabla.get_model()
        notvalid=[]
        model.foreach(self.delete_not_searched, (notvalid))
        notvalid.reverse()
        for host in notvalid:
            model.remove( model.get_iter(host) )
        
    def delete_not_searched(self, model, path, iter, data):
        txt=self.searchtxt.get_text()
        hostname=model.get_value(iter, 0)
        ip=model.get_value(iter, 1)
        username=model.get_value(iter, 2)
        if txt != hostname and txt != ip and txt != username:
            data.append(path)
  

    

    def exe_cmd(self, cmd):
        print_debug ( "exe_cmd() cmd=%s" %(cmd) )
        self.p = Popen(cmd, shell=True, bufsize=0, stdout=PIPE, stderr=STDOUT)
    
    def write_into_statusbar(self, msg, *args):
        #print_debug("STATUSBAR: Writing \"%s\" into statusbar" % msg)
        context_id=self.statusbar.get_context_id("status")
        self.statusbar.pop(context_id)
        self.statusbar.push(context_id, msg)
        return    

    def salirse(self,True):
        print_debug ( _("Exiting") )
        #gtk.main_quit()
        self.mainloop.quit()


    def run (self):
        self.mainloop = gobject.MainLoop()
        #self.mainloop.run()
        try:
            #gtk.main()
            self.mainloop.run()
        except KeyboardInterrupt: # Por si se pulsa Ctrl+C
            self.salirse(True)
    
"""
class thread_controller(Thread):
    # based on obex_controler
    # http://www.student.lu.se/~cif04usv/wiki/btrcv.html
    def __init__(self, cmd, gui):
        self.gui=gui
        self.finished=False
        Thread.__init__(self)
        self.p = Popen(cmd, shell=True, bufsize=0, stdout=PIPE, stderr=STDOUT)
        self.stdout = self.p.stdout
        #print_debug ( "ThreadController::__init__() Started %s with pid %d" % (cmd, self.p.pid ) )

    def isfinish(self):
        return self.finished
    
    def run(self):
        while True:
            #print "THREAD: Sleeping..."
            time.sleep(0.2)
            line = self.stdout.readline()
            line = line.replace('\n', '')
            # chek if terminated
            if self.p.poll() != None:
                self.finished=True

            if line.find("DONE") > 0:
                print_debug ( "ThreadController::run() DONE found" )
            #print("THREAD: read \" %s \" " % (line) )

            while gtk.events_pending():
                gtk.main_iteration(False)

            #self.gui.writeintoprogresstxt( line )
            return line
"""

if __name__ == '__main__':
    app = TcosMonitor ()
    # Run app
    app.run ()
